sdk/azidentity/device_code_credential.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azidentity
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+)
+
+const (
+	deviceCodeGrantType = "urn:ietf:params:oauth:grant-type:device_code"
+)
+
+// DeviceCodeCredential authenticates a user using the device code flow, and provides access tokens for that user account.
+// For more information on the device code authentication flow see https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code
+type DeviceCodeCredential struct {
+	client       *aadIdentityClient
+	tenantID     string       // Gets the Azure Active Directory tenant (directory) Id of the service principal
+	clientID     string       // Gets the client (application) ID of the service principal
+	callback     func(string) // Sends the user a message with a verification URL and device code to sign in to the login server
+	refreshToken string       // Gets the refresh token sent from the service and will be used to retreive new access tokens after the initial request for a token. Thread safety for updates is handled in the AuthenticationPolicy since only one goroutine will be updating at a time
+}
+
+// NewDeviceCodeCredential constructs a new DeviceCodeCredential with the details needed to authenticate against Azure Active Directory with a device code.
+// tenantID: The Azure Active Directory tenant (directory) ID of the service principal. If none is set then the default value ("organizations") will be used in place of the tenantID.
+// clientID: The client (application) ID of the service principal.
+// callback: The callback function used to send the login message back to the user
+// options: allow to configure the management of the requests sent to the Azure Active Directory service.
+func NewDeviceCodeCredential(tenantID string, clientID string, callback func(string), options *TokenCredentialOptions) *DeviceCodeCredential {
+	return &DeviceCodeCredential{tenantID: tenantID, clientID: clientID, callback: callback, client: newAADIdentityClient(options)}
+}
+
+// GetToken obtains a token from the Azure Active Directory service, following the device code authentication
+// flow. This function first requests a device code and requests that the user login to continue authenticating.
+// This function will keep polling the service for a token meanwhile the user logs.
+// scopes: The list of scopes for which the token will have access. The "offline_access" scope is checked for and automatically added in case it isn't present to allow for silent token refresh.
+// ctx: controlling the request lifetime.
+// Returns an AccessToken which can be used to authenticate service client calls.
+func (c *DeviceCodeCredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
+	for i, scope := range opts.Scopes {
```

Comment:
This block can be removed now yes?

Reply:
Oh I thought this had been moved.  It probably makes more sense to do this in the constructor so that it only happens once and not on every call to `GetToken()`.

Reply:
We dont know the scopes in the constructor though.

Reply:
Sigh yes you're right (I hate this scopes problem).